lib/spack/spack/test/build_environment.py

```diff
@@ -119,79 +130,48 @@ def _set_wrong_cc(x):
     assert os.environ['ANOTHER_VAR'] == 'THIS_IS_SET'
 
 
-@pytest.mark.usefixtures('config', 'mock_packages')
-def test_compiler_config_modifications(monkeypatch, working_env):
-    s = spack.spec.Spec('cmake')
-    s.concretize()
-    pkg = s.package
-
-    os.environ['SOME_VAR_STR'] = ''
-    os.environ['SOME_VAR_NUM'] = '0'
-    os.environ['PATH_LIST'] = '/path/third:/path/forth'
-    os.environ['EMPTY_PATH_LIST'] = ''
-    os.environ.pop('NEW_PATH_LIST', None)
-
-    env_mod = syaml_dict()
-    set_cmd = syaml_dict()
-    env_mod[syaml_str('set')] = set_cmd
-
-    set_cmd[syaml_str('SOME_VAR_STR')] = syaml_str('SOME_STR')
-    set_cmd[syaml_str('SOME_VAR_NUM')] = 1
-
-    monkeypatch.setattr(pkg.compiler, 'environment', env_mod)
-    spack.build_environment.setup_package(pkg, False)
-    assert os.environ['SOME_VAR_STR'] == 'SOME_STR'
-    assert os.environ['SOME_VAR_NUM'] == str(1)
-
-    env_mod = syaml_dict()
-    unset_cmd = syaml_dict()
-    env_mod[syaml_str('unset')] = unset_cmd
-
-    unset_cmd[syaml_str('SOME_VAR_STR')] = None
-
-    monkeypatch.setattr(pkg.compiler, 'environment', env_mod)
-    assert 'SOME_VAR_STR' in os.environ
-    spack.build_environment.setup_package(pkg, False)
-    assert 'SOME_VAR_STR' not in os.environ
-
-    env_mod = syaml_dict()
-    set_cmd = syaml_dict()
-    env_mod[syaml_str('set')] = set_cmd
-    append_cmd = syaml_dict()
-    env_mod[syaml_str('append-path')] = append_cmd
-    unset_cmd = syaml_dict()
-    env_mod[syaml_str('unset')] = unset_cmd
-    prepend_cmd = syaml_dict()
-    env_mod[syaml_str('prepend-path')] = prepend_cmd
-
-    set_cmd[syaml_str('EMPTY_PATH_LIST')] = syaml_str('/path/middle')
-
-    append_cmd[syaml_str('PATH_LIST')] = syaml_str('/path/last')
-    append_cmd[syaml_str('EMPTY_PATH_LIST')] = syaml_str('/path/last')
-    append_cmd[syaml_str('NEW_PATH_LIST')] = syaml_str('/path/last')
-
-    unset_cmd[syaml_str('SOME_VAR_NUM')] = None
+@pytest.mark.parametrize('initial,modifications,expected', [
+    # Set and unset variables
+    ({'SOME_VAR_STR': '', 'SOME_VAR_NUM': '0'},
+     {'set': {'SOME_VAR_STR': 'SOME_STR', 'SOME_VAR_NUM': 1}},
+     {'SOME_VAR_STR': 'SOME_STR', 'SOME_VAR_NUM': '1'}),
+    ({'SOME_VAR_STR': ''},
+     {'unset': ['SOME_VAR_STR']},
+     {'SOME_VAR_STR': None}),
+    # Append and prepend to the same variable
+    ({'EMPTY_PATH_LIST': '/path/middle'},
+     {'prepend_path': {'EMPTY_PATH_LIST': '/path/first'},
+      'append_path': {'EMPTY_PATH_LIST': '/path/last'}},
+     {'EMPTY_PATH_LIST': '/path/first:/path/middle:/path/last'}),
+    # Append and prepend from empty variables
+    ({'EMPTY_PATH_LIST': '', 'SOME_VAR_STR': ''},
```

Comment:
Make one of these not present in the initial dict, so we ensure we're setting properly when it's not defined at all.

Reply:
Added a new test for this case